File: backend/app/api/departments.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from typing import List

from app.core.database import get_db
from app.models.employee import Department
from app.schemas.employee import DepartmentCreate, DepartmentResponse
from app.api.deps import get_current_user

router = APIRouter(prefix="/departments", tags=["Departments"])


@router.get("/", response_model=List[DepartmentResponse])
async def get_departments(
    db: AsyncSession = Depends(get_db),
    current_user = Depends(get_current_user)
):
    result = await db.execute(select(Department).order_by(Department.name))
    return result.scalars().all()


@router.post("/", response_model=DepartmentResponse, status_code=status.HTTP_201_CREATED)
async def create_department(
    dept_data: DepartmentCreate,
    db: AsyncSession = Depends(get_db),
    current_user = Depends(get_current_user)
):
    dept = Department(**dept_data.model_dump())
    db.add(dept)
    await db.commit()
    await db.refresh(dept)
    return dept


@router.delete("/{department_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_department(
    department_id: int,
    db: AsyncSession = Depends(get_db),
    current_user = Depends(get_current_user)
):
    dept = await db.get(Department, department_id)
    if not dept:
        raise HTTPException(status_code=404, detail="Department not found")
    
    await db.delete(dept)
    await db.commit()

File: frontend/src/api/departments.ts
import api from './client'

export interface Department {
  department_id: number
  name: string
  manager_id: number | null
}

export interface DepartmentCreate {
  name: string
  manager_id?: number
}

export const getDepartments = async (): Promise<Department[]> => {
  const response = await api.get<Department[]>('/departments/')
  return response.data
}

export const createDepartment = async (data: DepartmentCreate): Promise<Department> => {
  const response = await api.post<Department>('/departments/', data)
  return response.data
}

export const deleteDepartment = async (id: number): Promise<void> => {
  await api.delete(`/departments/${id}`)
}
